doc/specification.md

```diff
@@ -0,0 +1,52 @@
+# Test Specification
+
+This document describes how to use the JSON test specifications to write your own test executor for
+[`wasi-testsuite`](..). The test executor included in this project provides a completely-usable
+reference implementation, but projects with other requirements may want to run the tests in their
+own way (e.g., no Python dependency). The JSON test specifications provide a simple way to verify
+that the tests indeed passed.
+
+### Configuration
+
+Before executing anything, a test executor is expected to:
+- find all `*.wasm` and `*.wat` files in a given subdirectory &mdash; these are the _test cases_
+- for each test case, look for a `.json` file in the same directory matching the base name (e.g.,
+  `foo.json` for `foo.wasm`) &mdash; parse this _specification_
+- if no `.json` file is present, use a default specification; a (conceptual) default specification
+  would look like:
+
+  ```json
+  {
+    "args": [],
+    "dirs": [],
+    "env": {},
+    "exit_code": 0,
+    "stderr": "",
+    "stdout": ""
+  }
+  ```
+
+- if the specification is missing fields, use default values
```

Comment:
I probably need to add somewhere that the executor must remove any `*.cleanup` files it finds in the path.

Reply:
Actually, @alexcrichton brought up a good point [here](https://github.com/bytecodealliance/wasmtime/pull/6343#discussion_r1185541513) that reminded me that there are other options here than creating the test files in-tree:
 - we could expect the test executor to copy the Wasm file (and associated "stuff") to some temporary directory which gets removed when complete
 - we could expect the test executor to prefix all of the `dirs` entries with a temporary directory; this would mean that we would have to fix some tests that have hard-coded `args` that match `dirs`
 - etc.

I don't really have a strong preference but others may so feel free to let me know.

Reply:
I think it's an interesting idea (I don't think we have to copy the wasm file itself, probably just files/directories related to it - but I'd leave it to the implementation). My suggestion is to just update the doc according to current implementation; I've created a ticket to track it: https://github.com/WebAssembly/wasi-testsuite/issues/75

Reply:
@alexcrichton, do you have any strong opinions on this? I guess we can discuss more in #75. I do think that if we keep the `*.cleanup` scheme (I added a line documenting this) we should also update the `.gitignore` to ignore these files so that they don't continually make the Git state dirty.

Reply:
Sounds like a reasonable starting position to me 👍